win32ss/user/user32/controls/ghost.c

```diff
@@ -0,0 +1,582 @@
+                    szNotRespondingA, ARRAYSIZE(szNotRespondingA));
+        StringCbCatA(szTextA, sizeof(szTextA), szNotRespondingA);
+        SetWindowTextA(hwnd, szTextA);
+    }
+
+    hwndPrev = GetWindow(hwndTarget, GW_HWNDPREV);
+
+    // hide target
+    ShowWindowAsync(hwndTarget, SW_HIDE);
+
+    // shrink the ghost to zero size and insert
+    SetWindowPos(hwnd, hwndPrev, 0, 0, 0, 0, SWP_NOMOVE | SWP_NOACTIVATE | SWP_NOOWNERZORDER | SWP_DRAWFRAME);
+
+    // resume the position and size of ghost
+    MoveWindow(hwnd, rc.left, rc.top, rc.right - rc.left, rc.bottom - rc.top, TRUE);
+
+    // make ghost visible
+    if (IsWindowUnicode(hwnd))
+        SetWindowLongPtrW(hwnd, GWL_STYLE, style | WS_VISIBLE);
+    else
+        SetWindowLongPtrA(hwnd, GWL_STYLE, style | WS_VISIBLE);
+
+    // redraw
+    InvalidateRect(hwnd, NULL, TRUE);
+
+    // set user data
+    pData->hwndTarget = hwndTarget;
+    pData->hbm32bpp = hbm32bpp;
+    if (IsWindowUnicode(hwnd))
+        SetWindowLongPtrW(hwnd, GWLP_USERDATA, (LONG_PTR)pData);
+    else
+        SetWindowLongPtrA(hwnd, GWLP_USERDATA, (LONG_PTR)pData);
+
+    SetPropW(hwndTarget, L"GhostProp", hwnd);
+
+    SetTimer(hwnd, GHOST_TIMER_ID, GHOST_INTERVAL, NULL);
+
+    return TRUE;
+}
+
+static void
+Ghost_Unenchant(HWND hwnd, BOOL bDestroyTarget)
+{
+    DWORD pid;
+    HANDLE hProcess;
+    GHOST_DATA *pData;
+    RECT rc;
+
+    pData = Ghost_GetData(hwnd);
+    if (!pData)
+        return;
+
+    RemovePropW(pData->hwndTarget, L"GhostProp");
+
+    DeleteObject(pData->hbm32bpp);
+    pData->hbm32bpp = NULL;
+
+    if (bDestroyTarget)
+    {
+        GetWindowThreadProcessId(pData->hwndTarget, &pid);
+        hProcess = OpenProcess(PROCESS_TERMINATE, FALSE, pid);
+        if (hProcess)
+        {
+            TerminateProcess(hProcess, -1);
+            CloseHandle(hProcess);
+        }
+
+        DestroyWindow(hwnd);
+        ShowWindowAsync(pData->hwndTarget, SW_SHOWNOACTIVATE);
+
+        DestroyWindow(pData->hwndTarget);
+        pData->hwndTarget = NULL;
+    }
+    else
+    {
+        GetWindowRect(hwnd, &rc);
+        SetWindowPos(pData->hwndTarget, NULL, rc.left, rc.top, 0, 0,
+            SWP_NOSIZE | SWP_NOOWNERZORDER | SWP_NOACTIVATE | SWP_NOSENDCHANGING);
+
+        ShowWindowAsync(pData->hwndTarget, SW_SHOWNOACTIVATE);
+        DestroyWindow(hwnd);
+    }
+}
+
+static void
+Ghost_OnDraw(HWND hwnd, HDC hdc)
+{
+    BITMAP bm;
+    HDC hdcMem;
+    GHOST_DATA *pData = Ghost_GetData(hwnd);
+
+    if (!pData || !GetObject(pData->hbm32bpp, sizeof(bm), &bm))
+        return;
+
+    hdcMem = CreateCompatibleDC(hdc);
+    if (hdcMem)
+    {
+        HGDIOBJ hbmOld = SelectObject(hdcMem, pData->hbm32bpp);
+        {
+            BitBlt(hdc, 0, 0, bm.bmWidth, bm.bmHeight, hdcMem, 0, 0, SRCCOPY | CAPTUREBLT);
+        }
+        SelectObject(hdcMem, hbmOld);
+        DeleteDC(hdcMem);
+    }
+}
+
+static void
+Ghost_OnNCPaint(HWND hwnd, HRGN hrgn)
```

Comment:
This function is also called from `Ghost_OnEraseBkgnd`, is it allowed to call `WM_NCPAINT` in here directly?

Reply:
OK.